testing/unittests/loggingtests.cpp

```diff
@@ -0,0 +1,485 @@
+        hasServiceGroupPredicate("hasService-id-predicate", LGSTGetTransactionID, "bar", 2, 3);
+        hasServiceGroupPredicate("hasService-id-predicate", LGSTGetTransactionID, "baz", 0, 5);
+    }
+
+    void testHasService()
+    {
+        hasServiceGroup("hasService-seed", LGSTGetTransactionSeed, "", true);
+        hasServiceGroup("hasService-seed", LGSTGetTransactionSeed, "foo", true);
+        hasServiceGroup("hasService-seed", LGSTGetTransactionSeed, "bar", true);
+        hasServiceGroup("hasService-seed", LGSTGetTransactionSeed, "baz", false);
+        hasServiceGroup("hasService-update", LGSTUpdateLOG, "", true);
+        hasServiceGroup("hasService-update", LGSTUpdateLOG, "foo", true);
+        hasServiceGroup("hasService-update", LGSTUpdateLOG, "bar", true);
+        hasServiceGroup("hasService-update", LGSTUpdateLOG, "baz", false);
+        hasServiceGroup("hasService-id", LGSTGetTransactionID, "", true);
+        hasServiceGroup("hasService-id", LGSTGetTransactionID, "foo", true);
+        hasServiceGroup("hasService-id", LGSTGetTransactionID, "bar", true);
+        hasServiceGroup("hasService-id", LGSTGetTransactionID, "baz", false);
+    }
+
+    void testGetTransactionSeed()
+    {
+        getTransactionSeed("getTransactionSeed", "", true, "alpha", "Transaction Seed");
+        getTransactionSeed("getTransactionSeed", "foo", true, "beta", "Transaction Seed");
+        getTransactionSeed("getTransactionSeed", "bar", true, "gamma", "Transaction Seed");
+        getTransactionSeed("getTransactionSeed", "baz", false, "Seed", "Failed");
+    }
+
+    void testGetTransactionId()
+    {
+        getTransactionId("getTransactionId", "", true, "alpha1", "");
+        getTransactionId("getTransactionId", "foo", true, "beta1", "");
+        getTransactionId("getTransactionId", "bar", true, "gamma1", "");
+        getTransactionId("getTransactionId", "baz", false, "", "");
+    }
+
+    LoggingPredicateTests()
+        : m_loader(nullptr, nullptr, nullptr, nullptr)
+    {
+    }
+
+    ILoggingManager* createManager()
+    {
+        Owned<IPTree> config(createPTreeFromXMLString(managerText));
+        Owned<ILoggingManager> manager(m_loader.create(*config));
+
+        manager->init(config, "unittest");
+        return manager.getClear();
+    }
+
+    void hasServiceGroupPredicate(const char* test, LOGServiceType service, const char* group, unsigned expectedPass, unsigned expectedFail)
+    {
+        using Variants = std::vector<Linked<const IEspLogAgentVariant> >;
+        Variants pass;
+        Variants fail;
+        VariantGroupPredicate groupPred(group);
+        EspLogAgentVariantPredicate wrapper = [&](const IEspLogAgentVariant& variant) {
+            if (groupPred(variant))
+                pass.emplace_back(&variant);
+            else
+                fail.emplace_back(&variant);
+            return false;
+        };
+
+        if (!group)
+            group = "";
```

Comment:
Not really. Passing nullptr to a StringBuffer constructor (the effect of line 382) yields an empty string buffer. This test is a sanity check before using the value in street, which doesn't react well to NULL parameters.

Reply:
Why is it different from the hasServiceGroup()?

Reply:
If you're asking what the difference is between hasServiceGroupPredicate and hasServiceGroup, this function tests the predicate using hasService while hasServiceGroup tests hasService.

The lambda function that wraps the group predicate here causes hasService to fail by not allowing the group predicate's response to be seen. The wrapper always returns false. It is a test of the predicate itself because it enables me to confirm that, regardless of agent order, the predicate both passed and failed the correct variants.

In the other function, hasService is tested with the group predicate and establishes that hasService used the predicate correctly.

Reply:
Renamed methods may make this distinction more obvious.